File: server/src/db/queries.ts
import { randomUUID } from 'node:crypto';
import { getDb, getWriteDb } from './index';
import { extractDisplayContent } from './content';
import type {
  Repo,
  Workspace,
  WorkspaceResponse,
  Session,
  SessionMessage,
  MessageResponse,
} from '@conductor-companion/shared';

export function listWorkspaces(): WorkspaceResponse[] {
  const db = getDb();
  const workspaces = db
    .prepare('SELECT * FROM workspaces ORDER BY updated_at DESC')
    .all() as Workspace[];

  return workspaces.map((ws) => {
    const repo = ws.repository_id
      ? (db
          .prepare('SELECT * FROM repos WHERE id = ?')
          .get(ws.repository_id) as Repo | undefined)
      : undefined;

    const sessionCount = (
      db
        .prepare(
          'SELECT COUNT(*) as count FROM sessions WHERE workspace_id = ? AND (is_hidden IS NULL OR is_hidden = 0)'
        )
        .get(ws.id) as { count: number }
    ).count;

    let activeSessionStatus: string | null = null;
    if (ws.active_session_id) {
      const session = db
        .prepare('SELECT status FROM sessions WHERE id = ?')
        .get(ws.active_session_id) as { status: string | null } | undefined;
      activeSessionStatus = session?.status ?? null;
    }

    return {
      ...ws,
      repo_name: repo?.name ?? null,
      repo_remote_url: repo?.remote_url ?? null,
      active_session_status: activeSessionStatus,
      session_count: sessionCount,
    };
  });
}

export function getWorkspace(id: string): WorkspaceResponse | null {
  const db = getDb();
  const ws = db
    .prepare('SELECT * FROM workspaces WHERE id = ?')
    .get(id) as Workspace | undefined;

  if (!ws) return null;

  const repo = ws.repository_id
    ? (db
        .prepare('SELECT * FROM repos WHERE id = ?')
        .get(ws.repository_id) as Repo | undefined)
    : undefined;

  const sessionCount = (
    db
      .prepare(
        'SELECT COUNT(*) as count FROM sessions WHERE workspace_id = ? AND (is_hidden IS NULL OR is_hidden = 0)'
      )
      .get(ws.id) as { count: number }
  ).count;

  let activeSessionStatus: string | null = null;
  if (ws.active_session_id) {
    const session = db
      .prepare('SELECT status FROM sessions WHERE id = ?')
      .get(ws.active_session_id) as { status: string | null } | undefined;
    activeSessionStatus = session?.status ?? null;
  }

  return {
    ...ws,
    repo_name: repo?.name ?? null,
    repo_remote_url: repo?.remote_url ?? null,
    active_session_status: activeSessionStatus,
    session_count: sessionCount,
  };
}

export function listSessions(workspaceId: string): Session[] {
  const db = getDb();
  return db
    .prepare(
      'SELECT * FROM sessions WHERE workspace_id = ? AND (is_hidden IS NULL OR is_hidden = 0) ORDER BY updated_at DESC'
    )
    .all(workspaceId) as Session[];
}

export function getSession(id: string): Session | null {
  const db = getDb();
  return (
    (db.prepare('SELECT * FROM sessions WHERE id = ?').get(id) as
      | Session
      | undefined) ?? null
  );
}

export function listMessages(
  sessionId: string,
  limit: number,
  before?: string
): MessageResponse[] {
  const db = getDb();

  let rows: SessionMessage[];
  if (before) {
    rows = db
      .prepare(
        'SELECT * FROM session_messages WHERE session_id = ? AND rowid < (SELECT rowid FROM session_messages WHERE id = ?) ORDER BY rowid DESC LIMIT ?'
      )
      .all(sessionId, before, limit) as SessionMessage[];
  } else {
    rows = db
      .prepare(
        'SELECT * FROM session_messages WHERE session_id = ? ORDER BY rowid DESC LIMIT ?'
      )
      .all(sessionId, limit) as SessionMessage[];
  }

  return rows.map((msg) => ({
    id: msg.id,
    session_id: msg.session_id,
    role: msg.role,
    display_content: extractDisplayContent(msg.content),
    created_at: msg.created_at,
    sent_at: msg.sent_at,
    model: msg.model,
    turn_id: msg.turn_id,
  }));
}

// --- Session with workspace path (for Claude CLI) ---

export interface SessionWithPath {
  session_id: string;
  claude_session_id: string | null;
  status: string | null;
  workspace_path: string | null;
}

export function getSessionWithWorkspacePath(sessionId: string): SessionWithPath | null {
  const db = getDb();
  const row = db
    .prepare(
      `SELECT
        s.id as session_id,
        s.claude_session_id,
        s.status,
        CASE
          WHEN r.root_path IS NOT NULL AND w.directory_name IS NOT NULL
          THEN REPLACE(r.root_path, '/repos/', '/workspaces/') || '/' || w.directory_name
          WHEN r.root_path IS NOT NULL
          THEN r.root_path
          ELSE NULL
        END as workspace_path
      FROM sessions s
      LEFT JOIN workspaces w ON s.workspace_id = w.id
      LEFT JOIN repos r ON w.repository_id = r.id
      WHERE s.id = ?`
    )
    .get(sessionId) as SessionWithPath | undefined;
  return row ?? null;
}

// --- Change detection queries ---

export function getLatestMessageRowid(): number {
  const db = getDb();
  const row = db
    .prepare('SELECT COALESCE(MAX(rowid), 0) as max_rowid FROM session_messages')
    .get() as { max_rowid: number };
  return row.max_rowid;
}

export function getMessagesAfterRowid(afterRowid: number): SessionMessage[] {
  const db = getDb();
  return db
    .prepare(
      'SELECT * FROM session_messages WHERE rowid > ? ORDER BY rowid ASC'
    )
    .all(afterRowid) as SessionMessage[];
}

export function getSessionsUpdatedAfter(after: string): Session[] {
  const db = getDb();
  return db
    .prepare(
      'SELECT * FROM sessions WHERE updated_at > ? ORDER BY updated_at ASC'
    )
    .all(after) as Session[];
}

export function getWorkspacesUpdatedAfter(after: string): Workspace[] {
  const db = getDb();
  return db
    .prepare(
      'SELECT * FROM workspaces WHERE updated_at > ? ORDER BY updated_at ASC'
    )
    .all(after) as Workspace[];
}

// --- Write operations (insert into Conductor's DB) ---

export function insertSessionMessage(
  sessionId: string,
  role: 'user' | 'assistant',
  content: string,
  turnId: string,
): string {
  const db = getWriteDb();
  const id = randomUUID();
  const now = new Date().toISOString();

  db.prepare(
    `INSERT INTO session_messages (id, session_id, role, content, created_at, sent_at, turn_id)
     VALUES (?, ?, ?, ?, ?, ?, ?)`
  ).run(id, sessionId, role, content, now, now, turnId);

  return id;
}

File: server-rust-archive/src/db/queries.rs
use sqlx::SqlitePool;

use super::models::*;

pub async fn list_repos(pool: &SqlitePool) -> anyhow::Result<Vec<Repo>> {
    let repos = sqlx::query_as::<_, Repo>("SELECT * FROM repos ORDER BY name")
        .fetch_all(pool)
        .await?;
    Ok(repos)
}

pub async fn list_workspaces(pool: &SqlitePool) -> anyhow::Result<Vec<WorkspaceResponse>> {
    let workspaces = sqlx::query_as::<_, Workspace>(
        "SELECT * FROM workspaces ORDER BY updated_at DESC",
    )
    .fetch_all(pool)
    .await?;

    let mut responses = Vec::new();
    for ws in workspaces {
        let repo = if let Some(ref repo_id) = ws.repository_id {
            sqlx::query_as::<_, Repo>("SELECT * FROM repos WHERE id = ?")
                .bind(repo_id)
                .fetch_optional(pool)
                .await?
        } else {
            None
        };

        let session_count = sqlx::query_scalar::<_, i64>(
            "SELECT COUNT(*) FROM sessions WHERE workspace_id = ? AND (is_hidden IS NULL OR is_hidden = 0)",
        )
        .bind(&ws.id)
        .fetch_one(pool)
        .await?;

        let active_session_status = if let Some(ref session_id) = ws.active_session_id {
            sqlx::query_scalar::<_, Option<String>>(
                "SELECT status FROM sessions WHERE id = ?",
            )
            .bind(session_id)
            .fetch_optional(pool)
            .await?
            .flatten()
        } else {
            None
        };

        responses.push(WorkspaceResponse {
            repo_name: repo.as_ref().and_then(|r| r.name.clone()),
            repo_remote_url: repo.as_ref().and_then(|r| r.remote_url.clone()),
            active_session_status,
            session_count,
            workspace: ws,
        });
    }

    Ok(responses)
}

pub async fn get_workspace(
    pool: &SqlitePool,
    id: &str,
) -> anyhow::Result<Option<WorkspaceResponse>> {
    let ws = sqlx::query_as::<_, Workspace>("SELECT * FROM workspaces WHERE id = ?")
        .bind(id)
        .fetch_optional(pool)
        .await?;

    let ws = match ws {
        Some(ws) => ws,
        None => return Ok(None),
    };

    let repo = if let Some(ref repo_id) = ws.repository_id {
        sqlx::query_as::<_, Repo>("SELECT * FROM repos WHERE id = ?")
            .bind(repo_id)
            .fetch_optional(pool)
            .await?
    } else {
        None
    };

    let session_count = sqlx::query_scalar::<_, i64>(
        "SELECT COUNT(*) FROM sessions WHERE workspace_id = ? AND (is_hidden IS NULL OR is_hidden = 0)",
    )
    .bind(&ws.id)
    .fetch_one(pool)
    .await?;

    let active_session_status = if let Some(ref session_id) = ws.active_session_id {
        sqlx::query_scalar::<_, Option<String>>("SELECT status FROM sessions WHERE id = ?")
            .bind(session_id)
            .fetch_optional(pool)
            .await?
            .flatten()
    } else {
        None
    };

    Ok(Some(WorkspaceResponse {
        repo_name: repo.as_ref().and_then(|r| r.name.clone()),
        repo_remote_url: repo.as_ref().and_then(|r| r.remote_url.clone()),
        active_session_status,
        session_count,
        workspace: ws,
    }))
}

pub async fn list_sessions(
    pool: &SqlitePool,
    workspace_id: &str,
) -> anyhow::Result<Vec<Session>> {
    let sessions = sqlx::query_as::<_, Session>(
        "SELECT * FROM sessions WHERE workspace_id = ? AND (is_hidden IS NULL OR is_hidden = 0) ORDER BY updated_at DESC",
    )
    .bind(workspace_id)
    .fetch_all(pool)
    .await?;
    Ok(sessions)
}

pub async fn get_session(pool: &SqlitePool, id: &str) -> anyhow::Result<Option<Session>> {
    let session = sqlx::query_as::<_, Session>("SELECT * FROM sessions WHERE id = ?")
        .bind(id)
        .fetch_optional(pool)
        .await?;
    Ok(session)
}

pub async fn list_messages(
    pool: &SqlitePool,
    session_id: &str,
    limit: i64,
    before: Option<&str>,
) -> anyhow::Result<Vec<SessionMessage>> {
    let messages = if let Some(before_id) = before {
        sqlx::query_as::<_, SessionMessage>(
            "SELECT * FROM session_messages WHERE session_id = ? AND rowid < (SELECT rowid FROM session_messages WHERE id = ?) ORDER BY rowid DESC LIMIT ?",
        )
        .bind(session_id)
        .bind(before_id)
        .bind(limit)
        .fetch_all(pool)
        .await?
    } else {
        sqlx::query_as::<_, SessionMessage>(
            "SELECT * FROM session_messages WHERE session_id = ? ORDER BY rowid DESC LIMIT ?",
        )
        .bind(session_id)
        .bind(limit)
        .fetch_all(pool)
        .await?
    };
    Ok(messages)
}

pub async fn get_latest_message_rowid(pool: &SqlitePool) -> anyhow::Result<i64> {
    let rowid = sqlx::query_scalar::<_, i64>(
        "SELECT COALESCE(MAX(rowid), 0) FROM session_messages",
    )
    .fetch_one(pool)
    .await?;
    Ok(rowid)
}

pub async fn get_messages_after_rowid(
    pool: &SqlitePool,
    after_rowid: i64,
) -> anyhow::Result<Vec<SessionMessage>> {
    let messages = sqlx::query_as::<_, SessionMessage>(
        "SELECT * FROM session_messages WHERE rowid > ? ORDER BY rowid ASC",
    )
    .bind(after_rowid)
    .fetch_all(pool)
    .await?;
    Ok(messages)
}

pub async fn get_sessions_updated_after(
    pool: &SqlitePool,
    after: &str,
) -> anyhow::Result<Vec<Session>> {
    let sessions = sqlx::query_as::<_, Session>(
        "SELECT * FROM sessions WHERE updated_at > ? ORDER BY updated_at ASC",
    )
    .bind(after)
    .fetch_all(pool)
    .await?;
    Ok(sessions)
}

pub async fn get_workspaces_updated_after(
    pool: &SqlitePool,
    after: &str,
) -> anyhow::Result<Vec<Workspace>> {
    let workspaces = sqlx::query_as::<_, Workspace>(
        "SELECT * FROM workspaces WHERE updated_at > ? ORDER BY updated_at ASC",
    )
    .bind(after)
    .fetch_all(pool)
    .await?;
    Ok(workspaces)
}
